File: 05b.js
import * as fs from 'fs';

function draw_line(x1, y1, x2, y2, buffer)
{
    var dx, dy;
    if (x1 === x2) {
        dx = 0;
    } else {
        dx = parseInt((x2 - x1) / Math.abs(x2 - x1)); // 1 or -1
    }
    if (y1 === y2) {
        dy = 0;
    } else {
        dy = parseInt((y2 - y1) / Math.abs(y2 - y1)); // 1 or -1
    }
    var x, y;
    [x, y] = [x1, y1];
    while (x != x2 || y != y2) {
        var key = String(x)+","+String(y);
        if (buffer[key] === undefined) {
            buffer[key] = 1;
        } else {
            buffer[key] += 1;
        }
        x += dx;
        y += dy;
    }
    var key = String(x)+","+String(y);
    if (buffer[key] === undefined) {
        buffer[key] = 1;
    } else {
        buffer[key] += 1;
    }
}

const filename = "05.dat";
fs.readFile(filename, (err, data) => {
    if (err) throw err;

    var lines = data.toString().split('\n');
    lines.pop();

    var buffer = {};
    lines.forEach(line => {
        var pairs = line.split(' -> ');
        var x1, y1;
        var x2, y2;
        [x1, y1] = pairs[0].split(',');
        [x2, y2] = pairs[1].split(',');
        draw_line(parseInt(x1), parseInt(y1), parseInt(x2), parseInt(y2), buffer);
    });
    var counter = 0;
    Object.keys(buffer).forEach(key => {
        if ((buffer[key]) > 1) {
            counter++;
        }
    });
    console.log(counter);
});

File: 05b.py
#!/usr/bin/python3

def draw_line(x1: int, y1: int, x2: int, y2: int, buffer: dict[(int, int)]) -> None:
    if x1 == x2:
        dx = 0
    else:
        dx = (x2 - x1) // abs(x2 - x1) # 1 or -1
    if y1 == y2:
        dy = 0
    else:
        dy = (y2 - y1) // abs(y2 - y1) # 1 or -1
    x, y = x1, y1
    while x != x2 or y != y2:
        try:
            buffer[(x, y)] += 1
        except KeyError:
            buffer[(x, y)] = 1
        x += dx
        y += dy
    try:
        buffer[(x, y)] += 1
    except KeyError:
        buffer[(x, y)] = 1

with open('05.dat') as f:
    data = f.read()
lines = data.split('\n')
lines.pop()

buffer = {}
for line in lines:
    pairs = line.split(' -> ')
    [x1, y1] = pairs[0].split(',')
    [x2, y2] = pairs[1].split(',')
    draw_line(int(x1), int(y1), int(x2), int(y2), buffer)
counter = 0
for key in buffer:
    if (buffer[key]) > 1:
        counter += 1
print(counter)